modules/auxiliary/server/capture/smtp.rb

```diff
@@ -40,91 +48,199 @@ def setup
   end
 
   def run
-    exploit()
+    exploit
   end
 
-  def on_client_connect(c)
-    @state[c] = {:name => "#{c.peerhost}:#{c.peerport}", :ip => c.peerhost, :port => c.peerport, :user => nil, :pass => nil}
-    c.put "220 SMTP Server Ready\r\n"
+  def auth_plain_parser(data)
+    # this data is \00 delimited, and has 3 fields: un\00un\00\pass.  Not sure why a double username, but we drop the first one
+    Rex::Text.decode_base64(data).split("\00").drop(1)
   end
```

Comment:
I haven't tested thoroughly, but something goes wrong in the plain auth parsing when the client sends a blank password.

```ruby
Net::SMTP.start('127.0.0.1', 25, 'localhost', 'username_plain', '', :plain)
```

This results in `AUTH PLAIN AHVzZXJuYW1lX3BsYWluAA==`, which decodes to:

```
# echo -n 'AHVzZXJuYW1lX3BsYWluAA==' | base64 -d
username_plain
```

Here's the output. Note that `username_plain` is stored as both the username and the password.

```
msf5 auxiliary(server/capture/smtp) > 
[*] SMTP: 127.0.0.1:36600 Command: EHLO localhost
[*] SMTP: 127.0.0.1:36600 Command: AUTH PLAIN AHVzZXJuYW1lX3BsYWluAA==
[+] SMTP LOGIN 127.0.0.1:36600 username_plain / username_plain
[*] SMTP: 127.0.0.1:36602 Command: EHLO localhost
[*] SMTP: 127.0.0.1:36602 Command: AUTH LOGIN
[*] SMTP: 127.0.0.1:36602 Command: dXNlcm5hbWVfbG9naW4=
[*] SMTP: 127.0.0.1:36602 Command: cGFzc3dvcmRfbG9naW4=
[+] SMTP LOGIN 127.0.0.1:36602 username_login / password_login
[*] SMTP: 127.0.0.1:36604 Command: EHLO localhost
[*] SMTP: 127.0.0.1:36604 Command: AUTH CRAM-MD5
[*] SMTP: 127.0.0.1:36604 Command: dXNlcm5hbWVfY3JhbSBhMDFiMWMxNzQyMmZmYzQwNjVmYTQ0MzE4NTVjODRkOQ==
[+] SMTP LOGIN 127.0.0.1:36604 username_cram / <58794434465@127.0.0.1>#a01b1c17422ffc4065fa4431855c84d9
[*] SMTP: 127.0.0.1:36606 Command: EHLO localhost
[*] SMTP: 127.0.0.1:36606 Command: MAIL FROM:<from@test.com>
[*] SMTP: 127.0.0.1:36606 Command: RCPT TO:<to@test.com>
[*] SMTP: 127.0.0.1:36606 Command: DATA
[*] SMTP: 127.0.0.1:36606 Command: test
.
[*] SMTP: 127.0.0.1:36606 EMAIL: test
[*] SMTP: 127.0.0.1:36606 Command: QUIT
[*] SMTP: 127.0.0.1:36608 Command: EHLO localhost
[*] SMTP: 127.0.0.1:36608 Command: MAIL FROM:<from@test.com>
[*] SMTP: 127.0.0.1:36608 Command: MAIL TO:<to@test.com>
[*] SMTP: 127.0.0.1:36608 Command: DATA
[*] SMTP: 127.0.0.1:36608 Command: RSET
[*] SMTP: 127.0.0.1:36610 Command: EHLO localhost
[*] SMTP: 127.0.0.1:36610 Command: MAIL FROM:<from@test.com>
[*] SMTP: 127.0.0.1:36610 Command: MAIL TO:<to@test.com>
[*] SMTP: 127.0.0.1:36610 Command: DATA
[*] SMTP: 127.0.0.1:36610 Command: testing a message which gets cancelled
RSET
[*] SMTP: 127.0.0.1:36610 EMAIL: testing a message which gets cancelled

msf5 auxiliary(server/capture/smtp) > creds
Credentials
===========

host       origin     service        public          private                                                   realm  private_type        JtR Format
----       ------     -------        ------          -------                                                   -----  ------------        ----------
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_login  password_login                                                   Password            
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_plain  username_plain                                                   Password            
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_cram   <58794434465@127.0.0.1>#a01b1c17422ffc4065fa4431855c84d9         Nonreplayable hash  hmac-md5

msf5 auxiliary(server/capture/smtp) > 
```

I don't think that's correct. It works fine when the username is blank.

```
msf5 auxiliary(server/capture/smtp) > creds
Credentials
===========

host       origin     service        public          private                                                  realm  private_type        JtR Format
----       ------     -------        ------          -------                                                  -----  ------------        ----------
127.0.0.1  127.0.0.1  25/tcp (smtp)                  password_plain                                                  Password            
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_login  password_login                                                  Password            
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_cram   <6914752990@127.0.0.1>#62f6bee180610b9a236102d83ea83808         Nonreplayable hash  hmac-md5

msf5 auxiliary(server/capture/smtp) > 
```

Comparatively, the other auth types work fine when the password is blank:

```
msf5 auxiliary(server/capture/smtp) > creds
Credentials
===========

host       origin     service        public          private                                                 realm  private_type        JtR Format
----       ------     -------        ------          -------                                                 -----  ------------        ----------
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_plain  username_plain                                                 Password            
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_login                                                                 Blank password      
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_cram   <745960849@127.0.0.1>#7b4d6ee99588a48b8ea0d2794c561dc6         Nonreplayable hash  hmac-md5
```

Reply:
This was actually a really interesting case.  If no username is submitted, the base64 decoded value is `\00username\00`, which meant the split was giving an array 1 shorter than expected, then the `.first` and `.last1` pulled the same value.  Good catch on this!

Reply:
Tested auth plain with username only and password only. Fixed.

```
Credentials
===========

host       origin     service        public          private                                                  realm  private_type        JtR Format
----       ------     -------        ------          -------                                                  -----  ------------        ----------
127.0.0.1  127.0.0.1  25/tcp (smtp)  asdf                                                                            Blank password      
127.0.0.1  127.0.0.1  25/tcp (smtp)                  asdf                                                            Password            
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_login  my password                                                     Password            
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_cram   <48677123003@0.0.0.0>#2c9526e9b93ec8573d4008230ede1c3f          Nonreplayable hash  hmac-md5
127.0.0.1  127.0.0.1  25/tcp (smtp)  username_cram   <678273648676@0.0.0.0>#109f033a54a6f2305190ed22eb95168d         Nonreplayable hash  hmac-md5
```